### PR TITLE
fix: Create new boundary to run onExit nodes

### PR DESCRIPTION
**Pull request by simster7**
Fixes: https://github.com/argoproj/argo-workflows/issues/5463

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -2937,10 +2937,10 @@ func (woc *wfOperationCtx) createTemplateContext(scope wfv1.ResourceScope, resou
 	}
 }
 
-func (woc *wfOperationCtx) runOnExitNode(ctx context.Context, templateRef, parentDisplayName, parentNodeName, boundaryID string, tmplCtx *templateresolution.Context) (bool, *wfv1.NodeStatus, error) {
+func (woc *wfOperationCtx) runOnExitNode(ctx context.Context, templateRef, parentNodeName, boundaryID string, tmplCtx *templateresolution.Context) (bool, *wfv1.NodeStatus, error) {
 	if templateRef != "" && woc.GetShutdownStrategy().ShouldExecute(true) {
 		woc.log.Infof("Running OnExit handler: %s", templateRef)
-		onExitNodeName := common.GenerateOnExitNodeName(parentDisplayName)
```
Comment (simster7) [inline workflow/controller/operator.go:2943]: `parentDisplayName` is non unique, causing nodes with the same display name (e.g. when using `withParam`) to create onExit nodes of the same name. After the first onExit node is created, subsequent onExit nodes created by different parent nodes (but with the same display name) will fail creation as they already exist. Use `parentNodeName` which is always unique instead.

File: workflow/controller/steps.go [hunk 2]
```diff
@@ -260,7 +260,7 @@ func (woc *wfOperationCtx) executeStepGroup(ctx context.Context, stepGroup []wfv
 		if !childNode.Fulfilled() {
 			completed = false
 		} else if childNode.Completed() {
-			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.OnExit, step.Name, childNode.Name, stepsCtx.boundaryID, stepsCtx.tmplCtx)
+			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.OnExit, childNode.Name, childNodeID, stepsCtx.tmplCtx)
```
Comment (simster7) [inline workflow/controller/steps.go:263]: On exit nodes should run under the boundary of the node that called them (the "exited" node)

**Review by alexec — COMMENTED** [review]
what will happen to in-flight workflows at time of upgrade?

File: workflow/controller/steps.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/steps.go:263]: Before:
```yaml
STEP                  TEMPLATE  PODNAME                       DURATION  MESSAGE
 ✔ dag-diamond-wzqnc  diamond
 ├─✔ A                echo      dag-diamond-wzqnc-2374547985  6s
 ├─✔ A.onExit         oe        dag-diamond-wzqnc-1294891656  6s
 ├─✔ B                echo      dag-diamond-wzqnc-2324215128  6s
 ├─✔ C                echo      dag-diamond-wzqnc-2340992747  6s
 ├─✔ B.onExit         oe        dag-diamond-wzqnc-250094443   6s
 ├─✔ C.onExit         oe        dag-diamond-wzqnc-728915506   6s
 ├─✔ D                echo      dag-diamond-wzqnc-2424880842  6s
 └─✔ D.onExit         oe        dag-diamond-wzqnc-2186811989  6s
```

After:
```yaml
STEP                  TEMPLATE  PODNAME                       DURATION  MESSAGE
 ✔ dag-diamond-ztvr2  diamond
 ├─✔ A                echo      dag-diamond-ztvr2-3230871262  2s
 │ └─✔ onExit         oe        dag-diamond-ztvr2-1416350393  6s
 ├─✔ B                echo      dag-diamond-ztvr2-3214093643  6s
 │ └─✔ onExit         oe        dag-diamond-ztvr2-2959263954  6s
 ├─✔ C                echo      dag-diamond-ztvr2-3197316024  6s
 │ └─✔ onExit         oe        dag-diamond-ztvr2-3577844235  6s
 └─✔ D                echo      dag-diamond-ztvr2-3314759357  6s
   └─✔ onExit         oe        dag-diamond-ztvr2-2103494724  6s
```

**Comment by simster7** [comment]
> what will happen to in-flight workflows at time of upgrade?

They will not fail, but there is a tiny chance that they end up redoing work (e.g. an onExit node executed twice) if the onExit node with the old name has been executed but not completed. If it has executed and completed, then its parent node will be succeeded and Argo won't try to schedule a new onExit node.

Regardless, I don't believe we provide any guarantees that you will be able to upgrade while running workflows seamlessly.

File: workflow/controller/steps.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/steps.go:263]: Note this is not only for cosmetic reasons: boundary parallelism is the main reason for making this change

**Comment by simster7** [comment]
This PR currently fixes two issues in one:

* https://github.com/argoproj/argo-workflows/pull/5478#discussion_r598857819 is a clear bug and a clear fix
* https://github.com/argoproj/argo-workflows/pull/5478#discussion_r598858569 fixes undefined behavior and the fix has to take into account how we want that behavior to be defined.

Should I split into this into two PRs so that we can merge the first fix independently?

File: workflow/controller/operator.go (hunk 1 above)
Comment (simster7) [reply workflow/controller/operator.go]: This fix has been split to https://github.com/argoproj/argo-workflows/pull/5486

**Comment by simster7** [comment]
This is now fixed in [`v3.0.0`](https://github.com/argoproj/argo-workflows/releases/tag/v3.0.0)

**Pull request closed without merging**